### PR TITLE
[common] Fix the includes in files (1/N)

### PR DESCRIPTION
**Pull request by kunaltyagi**
Fix the headers as per Include-What-You-Use.

Forgot to mark the PR as draft. Sorry for the noise.

TODO in another PR: pcl/common folder

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Do we need the Eigen include in pcl_macros? Does the Success macro undef really help since in several cases, there's Eigen included without using the pcl_macro file.

Also, where's the X11 header included? Directly by us or by third party (eg: VTK, etc.)?

**Comment by taketwo** [comment]
According to [this](https://eigen.tuxfamily.org/bz/show_bug.cgi?id=253), the offending macro comes out of "X.h". This file we do not explicitly include. Most likely indeed VTK is responsible.

The workaround was added very recently: https://github.com/PointCloudLibrary/pcl/commit/ed7569e0107def62472a4d09f047704afea5f21d. Somehow during the review, we did not question why it landed in `pcl_macros.h` and not somewhere else. @aPonza do you still recall what was the problem? Is there something special about your setup? Any idea why we didn't observe this clash before?

**Comment by aPonza** [comment]
Eigen/Core ([Eigen/src/Core/util/Memory.h](https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/util/Memory.h#L777)) is where `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` is defined, so it was there to include what I use. Without the `#undef` the code won't compile due to the documented X11 inclusion (likely VTK, yes):

```
[...]

[ 36%] Linking CXX executable ../bin/pcl_add_gaussian_noise
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/common/float_image_utils.cpp.o
[ 37%] Built target pcl_add_gaussian_noise
Scanning dependencies of target pcl_compute_hausdorff
[ 37%] Building CXX object tools/CMakeFiles/pcl_compute_hausdorff.dir/compute_hausdorff.cpp.o
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/vtk/pcl_image_canvas_source_2d.cpp.o
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/vtk/pcl_context_item.cpp.o
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/vtk/vtkRenderWindowInteractorFix.cpp.o
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/pcl_painter2D.cpp.o
[ 37%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/pcl_plotter.cpp.o
[ 38%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/vtk/vtkVertexBufferObject.cxx.o
In file included from /usr/include/eigen3/Eigen/Core:294:0,
                 from /opt/libpcl/pcl/common/include/pcl/pcl_macros.h:75,
                 from /opt/libpcl/pcl/visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h:35,
                 from /opt/libpcl/pcl/visualization/src/vtk/vtkVertexBufferObject.cxx:15:
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:425:2: error: #error The preprocessor symbol 'Success' is defined, possibly by the X11 header file X.h
 #error The preprocessor symbol 'Success' is defined, possibly by the X11 header file X.h
  ^
[ 38%] Building CXX object visualization/CMakeFiles/pcl_visualization.dir/src/vtk/vtkVertexBufferObjectMapper.cxx.o
In file included from /usr/include/eigen3/Eigen/Core:294:0,
                 from /opt/libpcl/pcl/common/include/pcl/pcl_macros.h:75,
                 from /opt/libpcl/pcl/visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h:35,
                 from /opt/libpcl/pcl/visualization/src/vtk/vtkVertexBufferObjectMapper.cxx:18:
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:425:2: error: #error The preprocessor symbol 'Success' is defined, possibly by the X11 header file X.h
 #error The preprocessor symbol 'Success' is defined, possibly by the X11 header file X.h
  ^
In file included from /usr/include/X11/Xlib.h:44:0,
                 from /usr/include/GL/glx.h:30,
                 from /usr/include/vtk-6.2/vtkgl.h:28,
                 from /opt/libpcl/pcl/visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h:34,
                 from /opt/libpcl/pcl/visualization/src/vtk/vtkVertexBufferObject.cxx:15:
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:432:3: error: expected identifier before numeric constant
   Success = 0,        
   ^
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:432:3: error: expected ‘}’ before numeric constant
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:432:3: error: expected unqualified-id before numeric constant
In file included from /usr/include/eigen3/Eigen/Core:294:0,
                 from /opt/libpcl/pcl/common/include/pcl/pcl_macros.h:75,
                 from /opt/libpcl/pcl/visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h:35,
                 from /opt/libpcl/pcl/visualization/src/vtk/vtkVertexBufferObject.cxx:15:
/usr/include/eigen3/Eigen/src/Core/util/Constants.h:544:1: error: expected declaration before ‘}’ token
 } // end namespace Eigen
 ^
visualization/CMakeFiles/pcl_visualization.dir/build.make:309: recipe for target 'visualization/CMakeFiles/pcl_visualization.dir/src/vtk/vtkVertexBufferObject.cxx.o' failed
make[2]: *** [visualization/CMakeFiles/pcl_visualization.dir/src/vtk/vtkVertexBufferObject.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

**Comment by taketwo** [comment]
OK, so the inclusion belongs here. Regarding the compilation error, it probably could be workaround by moving the "pcl_macros.h" include higher:

https://github.com/PointCloudLibrary/pcl/blob/af3ce2530b7ae8ed083a3515168626c587a5bbcd/visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h#L27-L36

**Comment by kunaltyagi** [comment]
In that case, we can actually remove Eigen from `pcl_macros.h` and directly include it before `vtkgl.h`. That'd make macros much easier, remove the `pcl_exports.h` since then nvcc wouldn't have to compile those deprecated items in Eigen...

**Comment by aPonza** [comment]
> we can actually remove Eigen from `pcl_macros.h`

you'd miss the include what you use on `EIGEN_MAKE_ALIGNED_OPERATOR_NEW`, wouldn't you? Or you'd have transitive inclusions? Am I missing something?

How about extracting `PCL_MAKE_ALIGNED_OPERATOR_NEW` into a new header instead, so pcl_macros stays simple like you want? It would entail a batch rename of the `#include` for all the aligned allocation classes.

**Comment by taketwo** [comment]
> you'd miss the include what you use on EIGEN_MAKE_ALIGNED_OPERATOR_NEW, wouldn't you?

Yep, I also think so.

> How about extracting PCL_MAKE_ALIGNED_OPERATOR_NEW into a new header instead,

I was thinking about this as well. What about a `shared_ptr.h` header that includes this macro, typedef for shared pointers, and `make_shared` implementation?

**Comment by kunaltyagi** [comment]
Creating a new file for `PCL_MAKE_ALIGNED_OP_NEW` makes sense, but not including it in `shared_ptr` simple because shared_ptr has no relation to this operation.

The existing use of the macro is related to Eigen (which is no surprise) but what's surprising is that moving out the Eigen include doesn't cause compile error.

Since the usage actually requires knowing about Eigen, maybe moving into `eigen_base.h` or similar is a better option. The vector map macros also technically belong with this macro (based on the usage). This will clean up the code a bit as well as not increase the dependence tree.

PS: Bike-shedding about the header name is welcome 😄

**Comment by taketwo** [comment]
> but not including it in shared_ptr simple because shared_ptr has no relation to this operation.

Yet this operation has relation to `make_shared`, and I'd put `make_shared` into "shared_ptr.h".

But I also see your point with `eigen_base.h`

**Comment by aPonza** [comment]
> What about a `shared_ptr.h` header that includes this macro, typedef for shared pointers, and `make_shared` implementation?

Sergio was also [against macro and make_shared together](https://github.com/PointCloudLibrary/pcl/pull/3146#discussion_r292863700). Not sure about which other macros you're referring to, couldn't find them

**Comment by kunaltyagi** [comment]
> Not sure about the vector map macros you're referring to, couldn't find them

Those beauties are [here](https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/impl/point_types.hpp#L194)

> put make_shared into "shared_ptr.h".

I'm totally for this. 👍

**Comment by aPonza** [comment]
Wow! So lines 194--260?I think eigen_base.h makes definite sense, and it's itself used in point_types.h as well, it shouldn't be much of an issue then.

**Comment by SergioRAgostinho** [comment]
> PS: Bike-shedding about the header name is welcome smile

Rip-off std c++ and call it `memory.h`.

**Comment by kunaltyagi** [comment]
In response to https://github.com/PointCloudLibrary/pcl/pull/3654#issuecomment-586961703 I've updated the PR with the lib->3rd->std order of includes

**Comment by kunaltyagi** [comment]
This seems to be done, except: tutorial CI is red. Culprit is the `-D PCL_NO_RECOMPILE=ON`. Any ideas on how to tackle this?

Relevant error log:
```
In file included from /__w/1/s/common/include/pcl/impl/instantiate.hpp:58:0,
                 from /__w/1/s/search/src/octree.cpp:39:
/__w/1/s/common/include/pcl/impl/instantiate.hpp:66:16: error: expected constructor, destructor, or type conversion before 'PCL_INSTANTIATE_Octree'
   BOOST_PP_CAT(PCL_INSTANTIATE_, TEMPLATE)(POINT_TYPE)
                ^
/usr/include/boost/preprocessor/seq/for_each.hpp:49:57: note: in expansion of macro 'PCL_INSTANTIATE_IMPL'
 # define BOOST_PP_SEQ_FOR_EACH_M_I(r, macro, data, seq) macro(r, data, BOOST_PP_SEQ_HEAD(seq))
                                                         ^
/__w/1/s/search/src/octree.cpp:44:1: note: in expansion of macro 'PCL_INSTANTIATE'
 PCL_INSTANTIATE(Octree, PCL_XYZ_POINT_TYPES)
 ^

```

**Comment by SergioRAgostinho** [comment]
If `PCL_NO_PRECOMPILE=ON` is defined there's supposedly no way for the code to reach line 66 in instantiate.hpp:66 :/

https://github.com/PointCloudLibrary/pcl/blob/08c7694ddbfd7773640d0c4e5f825efc98d3f547/common/include/pcl/impl/instantiate.hpp#L49-L70

**Comment by kunaltyagi** [comment]
Why am I getting all the weird compile errors?

**Comment by kunaltyagi** [comment]
> If PCL_NO_PRECOMPILE=ON is defined there's supposedly no way for the code to reach line 66

Is it possible that the library is compiled with some options and the tutorial with some other?

File: common/include/pcl/impl/instantiate.hpp [hunk 1]
```diff
@@ -35,14 +35,8 @@
...
  *
  */
-#ifndef PCL_IMPL_INSTANTIATE_H_
-#define PCL_IMPL_INSTANTIATE_H_
 
-#ifdef __GNUC__
-#pragma GCC system_header 
-#endif
-
-#include <pcl/pcl_config.h>
```
Comment (taketwo) [inline common/include/pcl/impl/instantiate.hpp]: You removed this include and thus you don't get `PCL_NO_PRECOMPILE` defined.

It's a very happy coincidence we caught this problem in tutorials CI. IWUY is a treacherous tool.

Comment (kunaltyagi) [reply common/include/pcl/impl/instantiate.hpp]: That's a real sad thing. I dislike macros, and now I think I hate macros without values. Thanks for the catch

Comment (kunaltyagi) [reply common/include/pcl/impl/instantiate.hpp]: Looks like it worked. Nice magic @taketwo

**Comment by taketwo** [comment]
> now I think I hate macros without value

That's pure evil, yes. In my projects, I always define "config" macros to either 1 or 0. Then in the code check `#if HAVE_XYZ`. If I forgot to include the config header, then I get a compilation error. Safety!

**Comment by kunaltyagi** [comment]
Waiting for #3654

**Comment by kunaltyagi** [comment]
common and impl folders still left. But I think that an be another PR

**Comment by kunaltyagi** [comment]
Such fun. The headers aren't order independent

**Comment by kunaltyagi** [comment]
Not failing on GCC 9.2, and this is weird again

**Comment by mvieth** [comment]
I am currently looking a bit through older issues/pull requests and just came across this PR. It seems like I unknowingly replicated some of your work from here in my PR series to optimize header includes, and now there are a lot of merge conflicts for this PR. Sorry about that. Do you think there is still something from this PR we can use or does the master branch already contain the changes (more or less)?

**Comment by kunaltyagi** [comment]
I'll check this again.

**Comment by larshg** [comment]
@mvieth Should we close this one, as per your own comment earlier and perhaps, rerun IWYU on a subset, to avoid these large PRs?

**Comment by mvieth** [comment]
Yes, can be closed from my side. We can run IWYU again, but I think most unnecessary headers are already removed.

**Pull request closed without merging**